chain/chain/src/chain.rs

```diff
@@ -3831,32 +3923,64 @@ impl Chain {
 
     /// Function to check whether we need to create a new snapshot while processing the current block
     /// Note that this functions is called as a part of block preprocesing, so the head is not updated to current block
-    fn should_make_or_delete_snapshot(&mut self) -> Result<(bool, bool), Error> {
+    fn should_make_or_delete_snapshot(&mut self) -> Result<SnapshotAction, Error> {
         // head value is that of the previous block, i.e. curr_block.prev_hash
         let head = self.head()?;
         if head.prev_block_hash == CryptoHash::default() {
             // genesis block, do not snapshot
-            return Ok((false, false));
+            return Ok(SnapshotAction::None);
         }
 
         let is_epoch_boundary =
             self.epoch_manager.is_next_block_epoch_start(&head.last_block_hash)?;
         let will_shard_layout_change =
             self.epoch_manager.will_shard_layout_change(&head.last_block_hash)?;
+        let protocol_version = self.epoch_manager.get_epoch_protocol_version(&head.epoch_id)?;
+
         let tries = self.runtime_adapter.get_tries();
         let snapshot_config = tries.state_snapshot_config();
-        let make_snapshot = match snapshot_config.state_snapshot_type {
-            // For every epoch, we snapshot if the next block would be in a different epoch
-            StateSnapshotType::EveryEpoch => is_epoch_boundary,
+        match snapshot_config.state_snapshot_type {
+            // For every epoch, we snapshot if the next block is the state sync "sync_hash" block
```

Comment:
Here I didn't do that because after removing the block header argument to this function, now the current block being preprocessed is not saved as the HEAD. So if that block is the "sync_hash" block, we won't be able to detect it until the next block. That's why below when we call `get_current_epoch_sync_hash()`, and it's equal to `head.last_block_hash`, we return `head.prev_block_hash`. Whereas in the prev epoch sync case, we actually do have the luxury of checking whether the next block is they sync block even without passing the block header being preprocessed to this function, because we can just use `epoch_manager.is_next_block_epoch_start()`.

So by not using `get_sync_hash()` and instead writing out each case explicitly, we can preserve the old behavior for the old state sync method where we take a snapshot when preprocessing the sync hash. It's probably not a huge deal but it means less is different, and so not as much of a chance of some unforseen bug. WDYT though? We could just snapshot one block later in both cases, but I wanted to change as little as possible

Reply:
Ok I'll just leave it for now since we're probably going to need to change the snapshot logic anyway after the changes to flat storage during resharding